### ISSUE TITLE
[inductor] `torch.any` on a uint8 tensor returns bool under `torch.compile` (eager and `torch.all` keep uint8), so downstream uint8 arithmetic silently changes

### ISSUE THREAD

**Issue by cuiliaomei-beep** [pytorch/pytorch#198081]
### 🐛 Describe the bug

**Summary.** `torch.any(x)` and `x.any(dim)` on a uint8 tensor return uint8 in eager and under `backend="aot_eager"`, while the Inductor-compiled function returns bool (torch 2.14.0 and nightly, CPU and CUDA); `torch.all` on the same tensor keeps uint8 under every backend. A user sees the compiled result change dtype without any warning, and arithmetic on it changes value because the bool is promoted to int64 instead of wrapping in uint8 (`torch.any(x) * 200 + 100` is `44` in eager and `300` compiled). Because Inductor registers its own `reduce_any` lowering that casts to bool and returns the bool reduction rather than going through the `_refs.any` decomposition that restores uint8 (which `aten.all` still uses), only `any` loses the legacy uint8 result type.

```python
import torch

x = torch.tensor([[0, 3], [0, 0]], dtype=torch.uint8)
for name, f in {"torch.any(x)": lambda x: torch.any(x), "torch.all(x)": lambda x: torch.all(x), "x.any(0)": lambda x: x.any(0),
                "any(x) * 200 + 100": lambda x: torch.any(x) * 200 + 100}.items():
    e = f(x)
    row = [f"eager {e.tolist()} {e.dtype}"]
    for backend in ("aot_eager", "inductor"):
        torch._dynamo.reset()
        c = torch.compile(f, backend=backend)(x)
        row.append(f"{backend} {c.tolist()} {c.dtype}")
    print(f"{name:20s} " + " | ".join(row))
```

```
torch.any(x)         eager 1 torch.uint8 | aot_eager 1 torch.uint8 | inductor True torch.bool
torch.all(x)         eager 0 torch.uint8 | aot_eager 0 torch.uint8 | inductor 0 torch.uint8
x.any(0)             eager [0, 1] torch.uint8 | aot_eager [0, 1] torch.uint8 | inductor [False, True] torch.bool
any(x) * 200 + 100   eager 44 torch.uint8 | aot_eager 44 torch.uint8 | inductor 300 torch.int64
```

**Root cause and possible fix.** `torch/_refs/__init__.py` has both `any` and `all` convert the result back to uint8 when the input is uint8 ("Preserves uint8 -- probably a legacy mask thing"), and `aten.all` reaches Inductor through that decomposition. `aten.any` does not: `torch/_inductor/lowering.py` registers `reduce_any`, which does `x = to_dtype(x, torch.bool)` and returns `make_reduction("any")(x, ...)`, so the result is bool. Either drop the dedicated lowering and let `aten.any` decompose like `aten.all`, or make `reduce_any` cast the result back to uint8 when the input dtype is uint8, matching the eager kernel.

### Error logs

No error is raised; the compiled function returns `torch.bool` where eager returns `torch.uint8`, and the values of the dependent arithmetic differ as shown above (`44` vs `300`).

### Versions

```
PyTorch version: 2.14.0+cpu
OS: Microsoft Windows 11 Pro for Workstations (10.0.26200 64-bit)
MSVC: Microsoft (R) C/C++ Optimizing Compiler Version 19.44.35228 for x64
Python version: 3.14.7 (64-bit runtime)
Is CUDA available: False
CPU: 11th Gen Intel(R) Core(TM) i9-11900T @ 1.50GHz
[pip3] numpy==2.5.2
[pip3] torch==2.14.0+cpu
```

Same output on nightly 2.15.0.dev20260911+cpu, on Linux with 2.14.0+cpu, and on a Tesla T4 (`x` on `cuda`) with 2.14.0+cu130 and nightly 2.15.0.dev20260921+cu130.

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

**Comment by vikrambal** [comment]
Hello, I would love to work on this. I confirmed the root cause by taking a look at the code:
_refs.any already preserves uint8 (torch/_refs/__init__.py), but Inductor's reduce_any lowering (torch/_inductor/lowering.py) bypasses this decomposition and does not cast back. aten.all has no dedicated lowering; thus, it stays correct through the _refs.all decomposition's uint8 cast. 

Here is what I am planning for the fix: cast reduce_any's result back to uint8 when the input dtype was uint8 (mirroring what _refs.all already does). This is opposed to dropping the dedicated lowering, which I think could regress performance on the fast path. Will add a regression test alongside test_any in test/inductor/test_torchinductor.py. 

Please let me know if this direction looks good before I put together the code change - happy to change it to your liking.
